The Biological Observation Matrix (BIOM) format
===============================================

[![Join the chat at https://gitter.im/biocore/biom-format](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/biocore/biom-format?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge&utm_content=badge)

[![biom-format CI](https://github.com/biocore/biom-format/actions/workflows/python-package-conda.yml/badge.svg)](https://github.com/biocore/biom-format/actions/workflows/python-package-conda.yml)

The BIOM file format (canonically pronounced *biome*) is designed to be a general-use format for representing counts of observations (e.g., OTUs, KO categories, lipid types) in one or more biological samples (e.g., microbiome samples, genomes, metagenomes).

Further details can be found at http://biom-format.org.

Getting help
------------

To get help with biom, you should use the [biom](http://stackoverflow.com/questions/tagged/biom) tag on StackOverflow (SO), or post to the [QIIME 2 Forum](http://forum.qiime2.org). Before posting a question, check out SO's guide on how to [ask a question](http://stackoverflow.com/questions/how-to-ask). The biom-format developers regularly monitor the `biom` SO tag.
