<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>The biom file format: Version 2.1 &mdash; biom-format.org</title>
    
    <link rel="stylesheet" href="../../_static/haiku.css" type="text/css" />
    <link rel="stylesheet" href="../../_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '../../',
        VERSION:     '2.1.14',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="../../_static/jquery.js"></script>
    <script type="text/javascript" src="../../_static/underscore.js"></script>
    <script type="text/javascript" src="../../_static/doctools.js"></script>
    <script type="text/javascript" src="../../_static/copybutton.js"></script>
    <link rel="top" title="biom-format.org" href="../../index.html" />
    <link rel="up" title="The biom file format" href="../biom_format.html" />
    <link rel="next" title="Quick start" href="../quick_usage_examples.html" />
    <link rel="prev" title="The biom file format: Version 2.0" href="biom-2.0.html" /> 
  </head>
  <body>
  
  <a href="https://github.com/biocore/biom-format"><img
    style="position: absolute; top: 0; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_red_aa0000.png"
    alt="Fork me on GitHub"></a>

      <div class="header"><h1 class="heading"><a href="../../index.html">
          <span>biom-format.org</span></a></h1>
        <h2 class="heading"><span>The biom file format: Version 2.1</span></h2>
      </div>
      <div class="topnav">
      
        <p>
        «&#160;&#160;<a href="biom-2.0.html">The biom file format: Version 2.0</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="../../index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="../quick_usage_examples.html">Quick start</a>&#160;&#160;»
        </p>

      </div>
      <div class="content">
        
        
  <div class="section" id="the-biom-file-format-version-2-1">
<span id="biom-2-1"></span><h1>The biom file format: Version 2.1<a class="headerlink" href="#the-biom-file-format-version-2-1" title="Permalink to this headline">¶</a></h1>
<p>The <tt class="docutils literal"><span class="pre">biom</span></tt> format is based on <a class="reference external" href="http://www.hdfgroup.org">HDF5</a> to provide the overall structure for the format. HDF5 is a widely supported binary format with native parsers available within many programming languages.</p>
<p>Required top-level attributes:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>id                   : &lt;string or null&gt; a field that can be used to id a table (or null)
type                 : &lt;string&gt; Table type (a controlled vocabulary)
                       Acceptable values:
                        &quot;OTU table&quot;
                        &quot;Pathway table&quot;
                        &quot;Function table&quot;
                        &quot;Ortholog table&quot;
                        &quot;Gene table&quot;
                        &quot;Metabolite table&quot;
                        &quot;Taxon table&quot;
format-url           : &lt;url&gt; A string with a static URL providing format details
format-version       : &lt;tuple&gt; The version of the current biom format, major and minor
generated-by         : &lt;string&gt; Package and revision that built the table
creation-date        : &lt;datetime&gt; Date the table was built (ISO 8601 format)
shape                : &lt;list of ints&gt;, the number of rows and number of columns in data
nnz                  : &lt;int&gt; The number of non-zero elements in the table
</pre></div>
</div>
<p>Required groups:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>observation/               : The HDF5 group that contains observation specific information and an observation oriented view of the data
observation/matrix         : The HDF5 group that contains matrix data oriented for observation-wise operations (e.g., in compressed sparse row format)
observation/metadata       : The HDF5 group that contains observation specific metadata information
observation/group-metadata : The HDF5 group that contains observation specific group metadata information (e.g., phylogenetic tree)
sample/                    : The HDF5 group that contains sample specific information and a sample oriented data oriented view of the data
sample/matrix              : The HDF5 group that contains matrix data oriented for sample-wise operations (e.g., in compressed sparse column format)
sample/metadata            : The HDF5 group that contains sample specific metadata information
sample/group-metadata      : The HDF5 group that contains sample specific group metadata information (e.g., relationships between samples)
</pre></div>
</div>
<p>Required datasets:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>observation/ids            : &lt;string&gt; or &lt;variable length string&gt; A (N,) dataset of the observation IDs, where N is the total number of IDs
observation/matrix/data    : &lt;float64&gt; A (nnz,) dataset containing the actual matrix data
observation/matrix/indices : &lt;int32&gt; A (nnz,) dataset containing the column indices (e.g., maps into samples/ids)
observation/matrix/indptr  : &lt;int32&gt; A (M+1,) dataset containing the compressed row offsets
sample/ids                 : &lt;string&gt; or &lt;variable length string&gt; A (M,) dataset of the sample IDs, where M is the total number of IDs
sample/matrix/data         : &lt;float64&gt; A (nnz,) dataset containing the actual matrix data
sample/matrix/indices      : &lt;int32&gt; A (nnz,) dataset containing the row indices (e.g., maps into observation/ids)
sample/matrix/indptr       : &lt;int32&gt; A (N+1,) dataset containing the compressed column offsets
</pre></div>
</div>
<p>Under the <tt class="docutils literal"><span class="pre">observation/metadata</span></tt> and <tt class="docutils literal"><span class="pre">sample/metadata</span></tt> groups, the user can specify an arbitrary number of datasets that represents a metadata category for that axis. The expected structure for each of these metadata datasets is a list of atomic type objects (int, float, str, ...) where the index order of the list corresponds to the index order of the relevant axis IDs. Special complex metadata fields have been defined, and they are stored in a specific way. Currently, the available special metadata fields are:</p>
<div class="highlight-python"><div class="highlight"><pre><span></span>observation/metadata/taxonomy      : &lt;string&gt; or &lt;variable length string&gt; A (N, ?) dataset containing the taxonomy names assigned to the observation
observation/metadata/KEGG_Pathways : &lt;string&gt; or &lt;variable length string&gt; A (N, ?) dataset containing the KEGG Pathways assigned to the observation
observation/metadata/collapsed_ids : &lt;string&gt; or &lt;variable length string&gt; A (N, ?) dataset containing the observation ids of the original table that have been collapsed in the given observation
sample/metadata/collapsed_ids      : &lt;string&gt; or &lt;variable length string&gt; A (M, ?) dataset containing the sample ids of the original table that have been collapsed in the given sample
</pre></div>
</div>
<p>Under the <tt class="docutils literal"><span class="pre">observation/group-metadata</span></tt> and <tt class="docutils literal"><span class="pre">sample/group-metadata</span></tt> groups, the user can specify an arbitrary number of datasets that represents a relationship between the ids for that axis. The expected structure for each of these group metadata datasets is a single string or variable length string. Each of these datasets should have defined an attribute called <tt class="docutils literal"><span class="pre">data_type</span></tt>, which specifies how the string should be interpreted. One example of such group metadata dataset is <tt class="docutils literal"><span class="pre">observation/group-metadata/phylogeny</span></tt>, with the attribute <tt class="docutils literal"><span class="pre">observation/group-metadata/phylogeny.attrs['data_type']</span> <span class="pre">=</span> <span class="pre">&quot;newick&quot;</span></tt>, which stores a single string with the newick format of the phylogenetic tree for the observations.</p>
<div class="section" id="example-biom-files">
<h2>Example biom files<a class="headerlink" href="#example-biom-files" title="Permalink to this headline">¶</a></h2>
<p>Below is an example of a rich biom file before binary compression into HDF5. Example binary files are located in the <a class="reference external" href="https://github.com/biocore/biom-format/tree/master/examples">github repository</a>.</p>
<div class="section" id="biom-2-1-otu-table-in-the-hdf5-data-description-language-ddl">
<h3>BIOM 2.1 OTU table in the HDF5 data description language (DDL)<a class="headerlink" href="#biom-2-1-otu-table-in-the-hdf5-data-description-language-ddl" title="Permalink to this headline">¶</a></h3>
<div class="highlight-python"><div class="highlight"><pre><span></span>HDF5 &quot;examples/rich_sparse_otu_table_hdf5.biom&quot; {
GROUP &quot;/&quot; {
   ATTRIBUTE &quot;creation-date&quot; {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): &quot;2014-07-29T16:16:36.617320&quot;
      }
   }
   ATTRIBUTE &quot;format-url&quot; {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): &quot;http://biom-format.org&quot;
      }
   }
   ATTRIBUTE &quot;format-version&quot; {
      DATATYPE  H5T_STD_I64LE
      DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
      DATA {
      (0): 2, 1
      }
   }
   ATTRIBUTE &quot;generated-by&quot; {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): &quot;example&quot;
      }
   }
   ATTRIBUTE &quot;id&quot; {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): &quot;No Table ID&quot;
      }
   }
   ATTRIBUTE &quot;nnz&quot; {
      DATATYPE  H5T_STD_I64LE
      DATASPACE  SCALAR
      DATA {
      (0): 15
      }
   }
   ATTRIBUTE &quot;shape&quot; {
      DATATYPE  H5T_STD_I64LE
      DATASPACE  SIMPLE { ( 2 ) / ( 2 ) }
      DATA {
      (0): 5, 6
      }
   }
   ATTRIBUTE &quot;type&quot; {
      DATATYPE  H5T_STRING {
         STRSIZE H5T_VARIABLE;
         STRPAD H5T_STR_NULLTERM;
         CSET H5T_CSET_ASCII;
         CTYPE H5T_C_S1;
      }
      DATASPACE  SCALAR
      DATA {
      (0): &quot;otu table&quot;
      }
   }
   GROUP &quot;observation&quot; {
      GROUP &quot;group-metadata&quot; {
      }
      DATASET &quot;ids&quot; {
         DATATYPE  H5T_STRING {
            STRSIZE H5T_VARIABLE;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_ASCII;
            CTYPE H5T_C_S1;
         }
         DATASPACE  SIMPLE { ( 5 ) / ( 5 ) }
         DATA {
         (0): &quot;GG_OTU_1&quot;, &quot;GG_OTU_2&quot;, &quot;GG_OTU_3&quot;, &quot;GG_OTU_4&quot;, &quot;GG_OTU_5&quot;
         }
      }
      GROUP &quot;matrix&quot; {
         DATASET &quot;data&quot; {
            DATATYPE  H5T_IEEE_F64LE
            DATASPACE  SIMPLE { ( 15 ) / ( 15 ) }
            DATA {
            (0): 1, 5, 1, 2, 3, 1, 1, 4, 2, 2, 1, 1, 1, 1, 1
            }
         }
         DATASET &quot;indices&quot; {
            DATATYPE  H5T_STD_I32LE
            DATASPACE  SIMPLE { ( 15 ) / ( 15 ) }
            DATA {
            (0): 2, 0, 1, 3, 4, 5, 2, 3, 5, 0, 1, 2, 5, 1, 2
            }
         }
         DATASET &quot;indptr&quot; {
            DATATYPE  H5T_STD_I32LE
            DATASPACE  SIMPLE { ( 6 ) / ( 6 ) }
            DATA {
            (0): 0, 1, 6, 9, 13, 15
            }
         }
      }
      GROUP &quot;metadata&quot; {
         DATASET &quot;taxonomy&quot; {
            DATATYPE  H5T_STRING {
               STRSIZE H5T_VARIABLE;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_ASCII;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SIMPLE { ( 5, 7 ) / ( 5, 7 ) }
            DATA {
            (0,0): &quot;k__Bacteria&quot;, &quot;p__Proteobacteria&quot;,
            (0,2): &quot;c__Gammaproteobacteria&quot;, &quot;o__Enterobacteriales&quot;,
            (0,4): &quot;f__Enterobacteriaceae&quot;, &quot;g__Escherichia&quot;, &quot;s__&quot;,
            (1,0): &quot;k__Bacteria&quot;, &quot;p__Cyanobacteria&quot;, &quot;c__Nostocophycideae&quot;,
            (1,3): &quot;o__Nostocales&quot;, &quot;f__Nostocaceae&quot;, &quot;g__Dolichospermum&quot;,
            (1,6): &quot;s__&quot;,
            (2,0): &quot;k__Archaea&quot;, &quot;p__Euryarchaeota&quot;, &quot;c__Methanomicrobia&quot;,
            (2,3): &quot;o__Methanosarcinales&quot;, &quot;f__Methanosarcinaceae&quot;,
            (2,5): &quot;g__Methanosarcina&quot;, &quot;s__&quot;,
            (3,0): &quot;k__Bacteria&quot;, &quot;p__Firmicutes&quot;, &quot;c__Clostridia&quot;,
            (3,3): &quot;o__Halanaerobiales&quot;, &quot;f__Halanaerobiaceae&quot;,
            (3,5): &quot;g__Halanaerobium&quot;, &quot;s__Halanaerobiumsaccharolyticum&quot;,
            (4,0): &quot;k__Bacteria&quot;, &quot;p__Proteobacteria&quot;,
            (4,2): &quot;c__Gammaproteobacteria&quot;, &quot;o__Enterobacteriales&quot;,
            (4,4): &quot;f__Enterobacteriaceae&quot;, &quot;g__Escherichia&quot;, &quot;s__&quot;
            }
         }
      }
   }
   GROUP &quot;sample&quot; {
      GROUP &quot;group-metadata&quot; {
      }
      DATASET &quot;ids&quot; {
         DATATYPE  H5T_STRING {
            STRSIZE H5T_VARIABLE;
            STRPAD H5T_STR_NULLTERM;
            CSET H5T_CSET_ASCII;
            CTYPE H5T_C_S1;
         }
         DATASPACE  SIMPLE { ( 6 ) / ( 6 ) }
         DATA {
         (0): &quot;Sample1&quot;, &quot;Sample2&quot;, &quot;Sample3&quot;, &quot;Sample4&quot;, &quot;Sample5&quot;,
         (5): &quot;Sample6&quot;
         }
      }
      GROUP &quot;matrix&quot; {
         DATASET &quot;data&quot; {
            DATATYPE  H5T_IEEE_F64LE
            DATASPACE  SIMPLE { ( 15 ) / ( 15 ) }
            DATA {
            (0): 5, 2, 1, 1, 1, 1, 1, 1, 1, 2, 4, 3, 1, 2, 1
            }
         }
         DATASET &quot;indices&quot; {
            DATATYPE  H5T_STD_I32LE
            DATASPACE  SIMPLE { ( 15 ) / ( 15 ) }
            DATA {
            (0): 1, 3, 1, 3, 4, 0, 2, 3, 4, 1, 2, 1, 1, 2, 3
            }
         }
         DATASET &quot;indptr&quot; {
            DATATYPE  H5T_STD_I32LE
            DATASPACE  SIMPLE { ( 7 ) / ( 7 ) }
            DATA {
            (0): 0, 2, 5, 9, 11, 12, 15
            }
         }
      }
      GROUP &quot;metadata&quot; {
         DATASET &quot;BODY_SITE&quot; {
            DATATYPE  H5T_STRING {
               STRSIZE H5T_VARIABLE;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_UTF8;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SIMPLE { ( 6 ) / ( 6 ) }
            DATA {
            (0): &quot;gut&quot;, &quot;gut&quot;, &quot;gut&quot;, &quot;skin&quot;, &quot;skin&quot;, &quot;skin&quot;
            }
         }
         DATASET &quot;BarcodeSequence&quot; {
            DATATYPE  H5T_STRING {
               STRSIZE H5T_VARIABLE;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_UTF8;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SIMPLE { ( 6 ) / ( 6 ) }
            DATA {
            (0): &quot;CGCTTATCGAGA&quot;, &quot;CATACCAGTAGC&quot;, &quot;CTCTCTACCTGT&quot;,
            (3): &quot;CTCTCGGCCTGT&quot;, &quot;CTCTCTACCAAT&quot;, &quot;CTAACTACCAAT&quot;
            }
         }
         DATASET &quot;Description&quot; {
            DATATYPE  H5T_STRING {
               STRSIZE H5T_VARIABLE;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_UTF8;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SIMPLE { ( 6 ) / ( 6 ) }
            DATA {
            (0): &quot;human gut&quot;, &quot;human gut&quot;, &quot;human gut&quot;, &quot;human skin&quot;,
            (4): &quot;human skin&quot;, &quot;human skin&quot;
            }
         }
         DATASET &quot;LinkerPrimerSequence&quot; {
            DATATYPE  H5T_STRING {
               STRSIZE H5T_VARIABLE;
               STRPAD H5T_STR_NULLTERM;
               CSET H5T_CSET_UTF8;
               CTYPE H5T_C_S1;
            }
            DATASPACE  SIMPLE { ( 6 ) / ( 6 ) }
            DATA {
            (0): &quot;CATGCTGCCTCCCGTAGGAGT&quot;, &quot;CATGCTGCCTCCCGTAGGAGT&quot;,
            (2): &quot;CATGCTGCCTCCCGTAGGAGT&quot;, &quot;CATGCTGCCTCCCGTAGGAGT&quot;,
            (4): &quot;CATGCTGCCTCCCGTAGGAGT&quot;, &quot;CATGCTGCCTCCCGTAGGAGT&quot;
            }
         }
      }
   }
}
}
</pre></div>
</div>
</div>
</div>
</div>


      </div>
      <div class="bottomnav">
      
        <p>
        «&#160;&#160;<a href="biom-2.0.html">The biom file format: Version 2.0</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="../../index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="../quick_usage_examples.html">Quick start</a>&#160;&#160;»
        </p>

      </div>


    <div class="footer">
        &copy; Copyright 2011-2022 The BIOM Format Development Team.
      Last updated on Jan 03, 2023.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.2.2.
    </div>
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-6636235-6']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>

  </body>
</html>