<!DOCTYPE html>

<html lang="en" data-content_root="../">
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" /><meta name="viewport" content="width=device-width, initial-scale=1" />

    <title>The biom file format &#8212; biom-format.org</title>
    <link rel="stylesheet" type="text/css" href="../_static/pygments.css?v=fa44fd50" />
    <link rel="stylesheet" type="text/css" href="../_static/haiku.css?v=dfa0e015" />
    <script src="../_static/documentation_options.js?v=30ffe591"></script>
    <script src="../_static/doctools.js?v=9a2dae69"></script>
    <script src="../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../_static/copybutton.js?v=1d1412a1"></script>
    <link rel="search" title="Search" href="../search.html" />
    <link rel="next" title="The biom file format: Version 1.0" href="format_versions/biom-1.0.html" />
    <link rel="prev" title="BIOM Documentation" href="index.html" /> 
  </head><body>
  
  <a href="https://github.com/biocore/biom-format"><img
    style="position: absolute; top: 0; right: 0; border: 0;"
    src="https://s3.amazonaws.com/github/ribbons/forkme_right_red_aa0000.png"
    alt="Fork me on GitHub"></a>

      <div class="header" role="banner"><h1 class="heading"><a href="../index.html">
          <span>biom-format.org</span></a></h1>
        <h2 class="heading"><span>The biom file format</span></h2>
      </div>
      <div class="topnav" role="navigation" aria-label="top navigation">
      
        <p>
        «&#160;&#160;<a href="index.html">BIOM Documentation</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="../index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="format_versions/biom-1.0.html">The biom file format: Version 1.0</a>&#160;&#160;»
        </p>

      </div>
      <div class="content" role="main">
        
        
  <section id="the-biom-file-format">
<span id="biom-format"></span><h1>The biom file format<a class="headerlink" href="#the-biom-file-format" title="Link to this heading">¶</a></h1>
<p>The BIOM project consists of two independent tools: the <cite>biom-format</cite> software package, which contains software tools for working with BIOM-formatted files and the tables they represent; and the BIOM file format. As of the 1.0.0 software version and the 1.0 file format version, the version of the software and the file format are independent of one another. Version specific documentation of the file formats can be found on the following pages.</p>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="format_versions/biom-1.0.html">The biom file format: Version 1.0</a><ul>
<li class="toctree-l2"><a class="reference internal" href="format_versions/biom-1.0.html#example-biom-files">Example biom files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="format_versions/biom-2.0.html">The biom file format: Version 2.0</a><ul>
<li class="toctree-l2"><a class="reference internal" href="format_versions/biom-2.0.html#example-biom-files">Example biom files</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="format_versions/biom-2.1.html">The biom file format: Version 2.1</a><ul>
<li class="toctree-l2"><a class="reference internal" href="format_versions/biom-2.1.html#example-biom-files">Example biom files</a></li>
</ul>
</li>
</ul>
</div>
<p>Release versions contain three integers in the following format: <code class="docutils literal notranslate"><span class="pre">major-version.minor-version.micro-version</span></code>. When <code class="docutils literal notranslate"><span class="pre">-dev</span></code> is appended to the end of a version string that indicates a development (or between-release version). For example, <code class="docutils literal notranslate"><span class="pre">1.0.0-dev</span></code> would refer to the development version following the 1.0.0 release.</p>
</section>
<section id="tips-and-faqs-regarding-the-biom-file-format">
<span id="sparse-or-dense"></span><h1>Tips and FAQs regarding the BIOM file format<a class="headerlink" href="#tips-and-faqs-regarding-the-biom-file-format" title="Link to this heading">¶</a></h1>
<section id="motivation-for-the-biom-format">
<h2>Motivation for the BIOM format<a class="headerlink" href="#motivation-for-the-biom-format" title="Link to this heading">¶</a></h2>
<p>The BIOM format was motivated by several goals. First, to facilitate efficient handling and storage of large, sparse biological contingency tables; second, to support encapsulation of core study data (contingency table data and sample/observation metadata) in a single file; and third, to facilitate the use of these tables between tools that support this format (e.g., passing of data between <a class="reference external" href="http://www.qiime.org">QIIME</a>, <a class="reference external" href="http://metagenomics.anl.gov">MG-RAST</a>, and <a class="reference external" href="http://vamps.mbl.edu/">VAMPS</a>.).</p>
<section id="efficient-handling-and-storage-of-very-large-tables">
<h3>Efficient handling and storage of very large tables<a class="headerlink" href="#efficient-handling-and-storage-of-very-large-tables" title="Link to this heading">¶</a></h3>
<p>In <a class="reference external" href="http://www.qiime.org">QIIME</a>, we began hitting limitations with OTU table objects when working with thousands of samples and hundreds of thousands of OTUs. In the near future we expect that we’ll be dealing with hundreds of thousands of samples in single analyses.</p>
<p>The OTU table format up to QIIME 1.4.0 involved a dense matrix: if an OTU was not observed in a given sample, that would be indicated with a zero. We now primarily represent OTU tables in a sparse format: if an OTU is not observed in a sample, there is no count for that OTU. The two ways of representing this data are exemplified here.</p>
<p>A dense representation of an OTU table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">OTU</span> <span class="n">ID</span> <span class="n">PC</span><span class="mf">.354</span>  <span class="n">PC</span><span class="mf">.355</span>  <span class="n">PC</span><span class="mf">.356</span>
<span class="n">OTU0</span>   <span class="mi">0</span>   <span class="mi">0</span>   <span class="mi">4</span>
<span class="n">OTU1</span>   <span class="mi">6</span>   <span class="mi">0</span>   <span class="mi">0</span>
<span class="n">OTU2</span>   <span class="mi">1</span>   <span class="mi">0</span>   <span class="mi">7</span>
<span class="n">OTU3</span>   <span class="mi">0</span>   <span class="mi">0</span>   <span class="mi">3</span>
</pre></div>
</div>
<p>A sparse representation of an OTU table:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">PC</span><span class="mf">.354</span> <span class="n">OTU1</span> <span class="mi">6</span>
<span class="n">PC</span><span class="mf">.354</span> <span class="n">OTU2</span> <span class="mi">1</span>
<span class="n">PC</span><span class="mf">.356</span> <span class="n">OTU0</span> <span class="mi">4</span>
<span class="n">PC</span><span class="mf">.356</span> <span class="n">OTU2</span> <span class="mi">7</span>
<span class="n">PC</span><span class="mf">.356</span> <span class="n">OTU3</span> <span class="mi">3</span>
</pre></div>
</div>
<p>OTU table data tends to be sparse (e.g., greater than 90% of counts are zero, and frequently as many as 99% of counts are zero) in which case the latter format is more convenient to work with as it has a smaller memory footprint. In biom-format 1.0.0, both of these representations are supported in the biom-format project via dense and sparse Table types. In biom-format 2.x, only sparse is supported as in practice, dense was not useful especially with improved study designs that utilize increasing numbers of samples.</p>
</section>
<section id="encapsulation-of-core-study-data-otu-table-data-and-sample-otu-metadata-in-a-single-file">
<h3>Encapsulation of core study data (OTU table data and sample/OTU metadata) in a single file<a class="headerlink" href="#encapsulation-of-core-study-data-otu-table-data-and-sample-otu-metadata-in-a-single-file" title="Link to this heading">¶</a></h3>
<p>Formats, such as JSON and HDF5, made more efficient storage of highly sparse data and allowed for storage of arbitrary amounts of sample and OTU metadata in a single file. Sample metadata corresponds to what is generally found in QIIME mapping files. At this stage inclusion of this information in the OTU table file is optional, but it may be useful for sharing these files with other QIIME users and for publishing or archiving results of analyses. OTU metadata (generally a taxonomic assignment for an OTU) is also optional. In contrast to the previous OTU table format, you can now store more than one OTU metadata value in this field, so for example you can score taxonomic assignments based on two different taxonomic assignment approaches.</p>
</section>
<section id="facilitating-the-use-of-tables-between-tools-that-support-this-format">
<h3>Facilitating the use of tables between tools that support this format<a class="headerlink" href="#facilitating-the-use-of-tables-between-tools-that-support-this-format" title="Link to this heading">¶</a></h3>
<p>Different tools, such as <a class="reference external" href="http://www.qiime.org">QIIME</a>, <a class="reference external" href="http://metagenomics.anl.gov">MG-RAST</a>, and <a class="reference external" href="http://vamps.mbl.edu/">VAMPS</a> work with similar data structures that represent different types of data. An example of this is a <cite>metagenome</cite> table that could be generated by MG-RAST (where for example, columns are metagenomes and rows are functional categories). Exporting this data from MG-RAST in a suitable format will allow for the application of many of the QIIME tools to this data (such as generation of alpha rarefaction plots or beta diversity ordination plots). This new format is far more general than previous formats, so will support adoption by groups working with different data types and is already being integrated to support transfer of data between <a class="reference external" href="http://www.qiime.org">QIIME</a>, <a class="reference external" href="http://metagenomics.anl.gov">MG-RAST</a>, and <a class="reference external" href="http://vamps.mbl.edu/">VAMPS</a>.</p>
</section>
</section>
<section id="file-extension">
<h2>File extension<a class="headerlink" href="#file-extension" title="Link to this heading">¶</a></h2>
<p>We recommend that BIOM files use the <code class="docutils literal notranslate"><span class="pre">.biom</span></code> extension.</p>
</section>
</section>


      </div>
      <div class="bottomnav" role="navigation" aria-label="bottom navigation">
      
        <p>
        «&#160;&#160;<a href="index.html">BIOM Documentation</a>
        &#160;&#160;::&#160;&#160;
        <a class="uplink" href="../index.html">Contents</a>
        &#160;&#160;::&#160;&#160;
        <a href="format_versions/biom-1.0.html">The biom file format: Version 1.0</a>&#160;&#160;»
        </p>

      </div>


    <div class="footer" role="contentinfo">
    &#169; Copyright 2011-2024 The BIOM Format Development Team.
      Last updated on May 11, 2024.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 7.3.7.
    </div>
<script type="text/javascript">

  var _gaq = _gaq || [];
  _gaq.push(['_setAccount', 'UA-6636235-6']);
  _gaq.push(['_trackPageview']);

  (function() {
    var ga = document.createElement('script'); ga.type = 'text/javascript'; ga.async = true;
    ga.src = ('https:' == document.location.protocol ? 'https://ssl' : 'http://www') + '.google-analytics.com/ga.js';
    var s = document.getElementsByTagName('script')[0]; s.parentNode.insertBefore(ga, s);
  })();

</script>

  </body>
</html>